### PR TITLE
Add SourcedText

### PR DESCRIPTION
**Pull request by csukuangfj**
## TODOs

- [x] Add tests for SourcedText

### CODE REVIEW

File: textsearch/csrc/utils.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * @param [in] row_ids   row_ids vector of length num_elems (
+ *                  row_ids[num_elems - 1] + 1 must equal num_rows). Must be
+ *                  non-decreasing.
+ * @param [in] num_rows   Number of rows in the irregular array, must
+ *                  be greater than any element of row_ids
+ * @param [out] row_splits  Row-splits vector that this function
+ *                  writes to, of length num_rows + 1.  row_splits[num_rows]
+ *                  will equal num_elems.
+ */
+void RowIdsToRowSplits(int32_t num_elems, const uint32_t *row_ids,
```
Comment (pzelasko) [inline textsearch/csrc/utils.h:24]: I'm not that familiar with this part of k2, could you explain to me what this function is doing / how row splits are going to be used in this project?

**Review by pzelasko — APPROVED** [review]

**Review by danpovey — APPROVED** [review]
Wow, this is great progress!  So it looks like you are implementing support for both the uint8 and int32 versions.
In order to get a 'minimum viable product', and allow us to make progress with other parts of the data preparation, I think it is not necessary to get the suffix-array stuff working properly; we can, initially, just do full Levenshtein alignment (with flexible begin and end in the reference), of the transcript to the reference text.  For this, the 'doc' part of SourcedText is not strictly necessary as each SourcedText will have just one document.  [If you want to  implemented it, the 'doc'  field could optionally be an int in these cases, to save space; but the danger of too many options is too much code to maintain.]

It would probably make sense to implement some basic normalization before doing the Levenshtein alignment, i.e. for purposes of the alignment we would normalize case and maybe remove punctuation characters.   You will have to figure out the best way to do this within SourcedText, while keeping track of the alignment information.  (Only removing punctuation characters would mess with the alignment information; normalizing text case would not.)

Later on, we can perhaps consider implementing versions of the suffix array construction algorithm that can handle larger alphabets.  For example, I found on google that "The SA-IS algorithm is one of the fastest known suffix array construction algorithms. A careful implementation by Yuta Mori outperforms most other linear or super-linear construction approaches; and it supports general alphabets where only character comparisons are allowed."  But this might not even be necessary.  If the number of distinct symbols in the text is not too large, and it won't be because it can't be more than the length of the input, we could modify CreateSuffixArray via some kind of wrapper or recursion to reduce K to the number of distinct symbols that are actually present in the text.  This would just use a std::unordered_set to find the distinct symbols, sort them, create a map from symbol to 'reordered' symbol, map the symbols in the input array, recurse to the main algorithm, then un-map the symbols and return.  That would avoid creating vast arrays in the RadixPass  part of the code.  But we can get into that later; it is only necessary if the speed of the simple Levenshtein stuff is not good enough due to references that are much longer than the audio (i.e. the book/chapter scenario).

File: textsearch/csrc/utils.h (hunk 1 above)
Comment (csukuangfj) [reply textsearch/csrc/utils.h:24]: Let me give you a concret example.

Suppose we have two doc:
  - Doc 0 contains "abc"
  - Doc 1 contains "defgh"

We store the content of all the doc into a single 1-D array. In the above example, the length of the array is
len(abc) + len(defgh) = 3 + 5 = 8.

## The purpose of row id
We need to answer the following question in O(1) time:

  - Which doc is the i-th element in the array from?

The `row_id` in our example is a 1-D array containing
```
0 0 0 1 1 1 1 1
```

- row_id[3] is 1, meaning element 3 in the array, which is `d`, is from doc 1. (note element index starts from 0)

We can know the following extra information from `row_id`:
  - Number of elements in the content array, which is `len(row_id)`, (i.e., 8 in our example)
  - Number of doc, which is `row_id[-1] + 1` (i.e., 2 in our example)
 

# The purpose of row splits

We want to know the following information in O(1) time

- How many elements does the i-th doc contain?
- Where is the first element of the i-th doc in the content array?

We can answer the above two questions by searching `row_id` but not in `O(1)` time.

The purpose of this function `RowIdsToRowSplits` is to compute `row_splits` from `row_ids`.
In our example, `row_ids` is 
```
0 0 0 1 1 1 1 1
```
and the computed `row_splits` would be
```
0 3 5
```

- `row_splits[0]` is 0, meaning the first element of doc 0 starts at position 0 in the content array.
- `row_splits[1]` is 3, meaning the first element of doc 1 starts at position 3 in the content array.

- `row_splits[1] - row_splits[0]` is the number of elements in doc 0
- `row_splits[2] - row_splits[1]` is the number of elements in doc 1
- `len(row_splits) - 1` is the number of doc

---

There is also a function `RowSplitsToRowIds` to compute `row_ids` from `row_splits` in `k2`. But we are not using that in `text_search`.

Comment (pzelasko) [reply textsearch/csrc/utils.h:24]: Thanks, this makes sense to me now!

**Comment by pzelasko** [comment]
For text normalization might be worth checking out Whisper for reference what they did; they have a set of "basic" rules (I assume for any language) here: https://github.com/openai/whisper/blob/main/whisper/normalizers/basic.py

As well as English-specific rules if that's useful: https://github.com/openai/whisper/blob/main/whisper/normalizers/english.py

**Pull request merged**